README
Net::Frame
==========

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Perl v5.6.1
  Bit::Vector
  Class::Gomor
  Net::IPv6Addr
  Socket6

GETTING HELP

A mailing list is available for all questions concerning Net::Frame::*.
Here is the link:
https://www.secure-side.com/lists/mailman/listinfo/netframe

COPYRIGHT AND LICENSE

You may distribute this module under the terms of the Artistic license.
See LICENSE.Artistic file in the source distribution archive.

Copyright (c) 2006-2014, Patrice <GomoR> Auffret



Changes
Revision history for Perl extension Net::Frame.

1.11 Tue Apr  8 15:33:43 CEST 2014
   - bugfix: https://rt.cpan.org/Public/Bug/Display.html?id=94035
   - bugfix: on making Socket6 optional, bugfix contributed by Andrew Feren
   - optimization: convertMac() contributed by Andrew Feren
   - optimization: inetChecksum() contributed by Andrew Feren

1.10 Mon Mar 10 13:26:53 CET 2014
   - UPDATE: makes Socket6 optional, contributed by Andrew Feren
   - update: copyright notice

1.09 Wed Jan 25 22:48:18 CET 2012
   - UPDATE: %Next in UDP and TCP layers are now empty by default. The
   developper who wants to add a new layer 7 seven will have to fill the
   variable straight from his module.

1.08 Sat Jan 14 09:49:09 CET 2012
   - bugfix: TCP and UDP checksum computation when using another
     Net::Frame::Layer object (instead of plain payload).
   => Thanks to Vince
   - update: copyright notice

1.07 Sun Feb 20 17:49:11 CET 2011
   - bugfix: in ARP pack/unpack
   - bugfix: AF_INET6 constant badly used
   - UPDATE: IGMPv4 renamed to IGMP
   - update: copyright notice

1.06 Sat May 23 15:33:44 CEST 2009
   - update: TCP computeChecksums()
   - update: UDP computeChecksums()
   - update: IPv4 added UDPLite support
   - update: IPv4 computeLengths()
   - update: copyright notice

1.05 Sun Nov  9 22:51:39 CET 2008
   - new: computeLengths() in ETH, to use 802.3 easily
   - new: $Next global variable in each module to allow customized encapsulation()
          => mainly used to encapsulate application layer procotols
   - new: UDP now have some $Next values (DHCP, RIPv1, HSRP)
   - new: added link to a MPLS layer from ETH and SLL layers
   - bugfix: pad TCP and UDP raw data if less than required length

1.04 Mon Mar 26 23:19:44 CEST 2007
   - bugfix: TCP: computeLengths() with options
   - update: ETH: more checks for next layer decoding (LLC)

1.03 Wed Jan  3 22:44:59 CET 2007
   - update: Net::Libdnet no more required

1.02 Thu Dec 21 23:51:07 CET 2006
   - bugfix: forgot to load some subs for IPv6 subs in Layer.pm

1.01 Sun Dec 17 17:45:32 CET 2006
   - UPDATE: moved all layers to Net::Frame::Layer::* namespace
   - bugfix: IPv6 layer handling in computeLengths() for TCP layer

1.00 Sat Dec  9 18:39:09 CET 2006
   - first public release



lib/Net/Frame/UDP.pm
#
# $Id: UDP.pm,v 1.9 2006/12/09 17:33:07 gomor Exp $
#
package Net::Frame::UDP;
use strict;
use warnings;

use Net::Frame::Layer qw(:consts :subs);
require Exporter;
our @ISA = qw(Net::Frame::Layer Exporter);

our %EXPORT_TAGS = (
   consts => [qw(
      NF_UDP_HDR_LEN
   )],
);
our @EXPORT_OK = (
   @{$EXPORT_TAGS{consts}},
);

use constant NF_UDP_HDR_LEN => 8;

our @AS = qw(
   src
   dst
   length
   checksum
);
__PACKAGE__->cgBuildIndices;
__PACKAGE__->cgBuildAccessorsScalar(\@AS);

no strict 'vars';

sub new {
   shift->SUPER::new(
      src      => getRandomHighPort(),
      dst      => 0,
      length   => 0,
      checksum => 0,
      @_,
   );
}

sub pack {
   my $self = shift;

   $self->[$__raw] = $self->SUPER::pack('nnnn',
      $self->[$__src],
      $self->[$__dst],
      $self->[$__length],
      $self->[$__checksum],
   ) or return undef;

   if ($self->[$__payload]) {
      $self->[$__raw] =
         $self->[$__raw].$self->SUPER::pack('a*', $self->[$__payload])
            or return undef;
   }

   $self->[$__raw];
}

sub unpack {
   my $self = shift;

   my ($src, $dst, $len, $checksum, $payload) =
      $self->SUPER::unpack('nnnn a*', $self->[$__raw])
         or return undef;

   $self->[$__src]      = $src;
   $self->[$__dst]      = $dst;
   $self->[$__length]   = $len;
   $self->[$__checksum] = $checksum;
   $self->[$__payload]  = $payload;

   $self;
}

sub getLength { NF_UDP_HDR_LEN }

sub getPayloadLength { shift->SUPER::getPayloadLength }

sub computeLengths {
   my $self = shift;
   $self->[$__length] = $self->getLength + $self->getPayloadLength;
   1;
}

sub computeChecksums {
   my $self = shift;
   my ($h)  = @_;

   my $phpkt;
   if ($h->{type} eq 'IPv4') {
      $phpkt = $self->SUPER::pack('a4a4CCn',
         inetAton($h->{src}), inetAton($h->{dst}), 0, 17, $self->[$__length],
      );
   }
   elsif ($h->{type} eq 'IPv6') {
      $phpkt = $self->SUPER::pack('a*a*NnCC',
         inet6Aton($h->{src}),
         inet6Aton($h->{dst}), $self->[$__length], 0, 0, 17,
      );
   }

   $phpkt .= $self->SUPER::pack('nnnn',
      $self->[$__src], $self->[$__dst], $self->[$__length], 0,
   ) or return undef;

   if ($self->[$__payload]) {
      $phpkt .= $self->SUPER::pack('a*', $self->[$__payload])
         or return undef;
   }

   $self->[$__checksum] = inetChecksum($phpkt);

   1;
}

sub encapsulate { shift->[$__nextLayer] }

sub getKey {
   my $self = shift;
   $self->layer.':'.$self->[$__src].'-'.$self->[$__dst];
}

sub getKeyReverse {
   my $self = shift;
   $self->layer.':'.$self->[$__dst].'-'.$self->[$__src];
}

sub match {
   my $self = shift;
   my ($with) = @_;
   1;
}

sub print {
   my $self = shift;

   my $l = $self->layer;
   sprintf
      "$l: src:%d  dst:%d  length:%d  checksum:0x%02x",
         $self->[$__src], $self->[$__dst], $self->[$__length],
         $self->[$__checksum];
}

1;

__END__

=head1 NAME

Net::Frame::UDP - User Datagram Protocol layer object

=head1 SYNOPSIS

   use Net::Frame::UDP qw(:consts);

   # Build a layer
   my $layer = Net::Frame::UDP->new(
      src      => getRandomHighPort(),
      dst      => 0,
      length   => 0,
      checksum => 0,
   );
   $layer->pack;

   print 'RAW: '.$layer->dump."\n";

   # Read a raw layer
   my $layer = Net::Frame::UDP->new(raw = $raw);

   print $layer->print."\n";
   print 'PAYLOAD: '.unpack('H*', $layer->payload)."\n"
      if $layer->payload;

=head1 DESCRIPTION

This modules implements the encoding and decoding of the UDP layer.

RFC: ftp://ftp.rfc-editor.org/in-notes/rfc768.txt

See also B<Net::Frame::Layer> for other attributes and methods.

=head1 ATTRIBUTES

=over 4

=item B<src>

=item B<dst>

Source and destination ports.

=item B<length>

The length in bytes of the datagram, including layer 7 payload (that is, layer 4 + layer 7).

=item B<checksum>

Checksum of the datagram.

=back

The following are inherited attributes. See B<Net::Frame::Layer> for more information.

=over 4

=item B<raw>

=item B<payload>

=item B<nextLayer>

=back

=head1 METHODS

=over 4

=item B<new>

=item B<new> (hash)

Object constructor. You can pass attributes that will overwrite default ones. See B<SYNOPSIS> for default values.

=item B<computeLengths>

Computes various lengths contained within this layer.

=item B<computeChecksums> ({ type => PROTO, src => IP, dst => IP })

In order to compute checksums of TCP, you need to pass the protocol type (IPv4, IPv6), the source and destination IP addresses (IPv4 for IPv4, IPv6 for IPv6).

=item B<getKey>

=item B<getKeyReverse>

These two methods are basically used to increase the speed when using B<recv> method from B<Net::Frame::Simple>. Usually, you write them when you need to write B<match> method.

=item B<match> (Net::Frame::UDP object)

This method is mostly used internally. You pass a B<Net::Frame::ARP> layer as a parameter, and it returns true if this is a response corresponding for the request, or returns false if not.

=back

The following are inherited methods. Some of them may be overriden in this layer, and some others may not be meaningful in this layer. See B<Net::Frame::Layer> for more information.

=over 4

=item B<layer>

=item B<computeLengths>

=item B<computeChecksums>

=item B<pack>

=item B<unpack>

=item B<encapsulate>

=item B<getLength>

=item B<getPayloadLength>

=item B<print>

=item B<dump>

=back

=head1 CONSTANTS

No constants here.

=head1 SEE ALSO

L<Net::Frame::Layer>

=head1 AUTHOR

Patrice E<lt>GomoRE<gt> Auffret

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2006, Patrice E<lt>GomoRE<gt> Auffret

You may distribute this module under the terms of the Artistic license.
See LICENSE.Artistic file in the source distribution archive.

=cut
